From 3ee85d67242fa038ab60ba1f4ee23a1d122fa415 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Thu, 29 Nov 2012 17:58:13 -0700
Subject: [PATCH] Display CHANGE dashboards on the ChangeScreen.

If a project has a CHANGE dashboard defined for it (either
directly or via inheritance), then any changes in that project
will display this dashboard in a disclosure panel (named after
the dashboard) on the ChangeScreen below the description block.
This means that a site can configure such a dashboard for every
change by simply defining one dashboard in All-Projects and
pointing the All-Projects config to it.  Or alternatively, if
this feature is not useful to a site, if they never point their
project config to a "change" dashboard then this feature will
simply not come into play.

Change-Id: Ie55861a7fa0f510e71b9ebb8edaf242fafad2e76
---
 Documentation/user-dashboards.txt                  |   16 ++++++++++++++++
 .../google/gerrit/client/changes/ChangeScreen.java |   16 ++++++++++++++++
 2 files changed, 32 insertions(+), 0 deletions(-)

diff --git a/Documentation/user-dashboards.txt b/Documentation/user-dashboards.txt
index c2bee92..2910491 100644
--- a/Documentation/user-dashboards.txt
+++ b/Documentation/user-dashboards.txt
@@ -155,6 +155,22 @@ but not inherited to the child projects.
   local-default = refs/meta/dashboards/main:local
 ----
 
+Change Dashboard
+----------------
+
+A change dashboard may be defined which will be displayed on each
+change in a project.  Similarly to defining a project default
+dashboard, the change dashboard will be inherited by child projects if
+they do not define their own.  Likewise it is possible to define a
+`local-change` that is only used by this project but not inherited to
+the child projects.
+
+----
+[dashboard]
+  change = refs/meta/dashboards/main:change
+  local-change = refs/meta/dashboards/main:local_change
+----
+
 GERRIT
 ------
 Part of link:index.html[Gerrit Code Review]
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
index e53beea..4a5b57a 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/changes/ChangeScreen.java
@@ -69,6 +69,7 @@ public class ChangeScreen extends Screen
   private ChangeTable dependencies;
   private ChangeTable.Section dependsOn;
   private ChangeTable.Section neededBy;
+  private FlowPanel dashboardPanel;
 
   private PatchSetsBlock patchSetsBlock;
 
@@ -168,6 +169,10 @@ public class ChangeScreen extends Screen
     descriptionBlock = new ChangeDescriptionBlock(keysAction);
     add(descriptionBlock);
 
+    dashboardPanel = new FlowPanel();
+    dashboardPanel.setVisible(false);
+    add(dashboardPanel);
+
     approvals = new ApprovalTable();
     add(approvals);
 
@@ -276,6 +281,17 @@ public class ChangeScreen extends Screen
       includedInPanel.setVisible(false);
     }
 
+    String params = detail.getDashboardParameters();
+    if (params != null) {
+      DisclosurePanel dp = new DisclosurePanel(
+          detail.getDashboardTitle());
+      DashboardTable dt = new DashboardTable(params);
+      dt.setWidth("99%");
+      dp.setContent(dt);
+      dashboardPanel.add(dp);
+      dashboardPanel.setVisible(true);
+    }
+
     dependencies.setAccountInfoCache(detail.getAccounts());
     approvals.setAccountInfoCache(detail.getAccounts());
 
-- 
1.7.8

